scipy/sparse/construct.py

```diff
@@ -664,16 +664,27 @@ def block_diag(mats, format=None, dtype=None):
            [0, 0, 0, 7]])
 
     """
-    nmat = len(mats)
-    rows = []
+    dtype = np.dtype(dtype)
+    row = []
+    col = []
+    data = []
+    r_idx = 0
+    c_idx = 0
     for ia, a in enumerate(mats):
-        row = [None]*nmat
         if issparse(a):
-            row[ia] = a
+            a = a.tocsr()
         else:
-            row[ia] = coo_matrix(a)
-        rows.append(row)
-    return bmat(rows, format=format, dtype=dtype)
+            a = coo_matrix(a).tocsr()
```

Comment:
What about `a = csr_matrix(a)`?

Reply:
But also, why force CSR format here?

Reply:
My intention there was to transform all input (sparse or dense arrays, lists, etc.) to the same type and format, so that I could handle them all in a uniform way afterwards. I chose CSR (over COO, for example) because it supports indexing. Do you think I should handle this in a cleaner way?

I agree with `a = csr_matrix(a)` :smile: